### PR TITLE
feat: add Messenger class for multiple adapter failover support

### PR DESCRIPTION
**Pull request by TorstenDittmann**
## Summary

This PR introduces the `Messenger` class, which enables automatic failover across multiple messaging adapters. This addresses the feature request to allow using multiple adapters where if one fails, it falls back to the next one.

## Changes

- **New Class:** `Utopia\Messaging\Messenger`
  - Orchestrates multiple adapters for failover support
  - Accepts a single `Adapter` or an array of `Adapter[]`
  - Tries adapters sequentially when exceptions are thrown
  - Validates all adapters are compatible (same type and message type)
  - Returns the first successful response
  - Throws aggregated exception with detailed error messages if all adapters fail

- **New Tests:** `tests/Messaging/MessengerTest.php`
  - 14 comprehensive test cases
  - Tests success, failover, and all-fail scenarios
  - Tests adapter validation and edge cases

- **Documentation:** Updated README.md with usage example

## Behavior

- **Failover trigger:** Only on exceptions (not on returned failure payloads)
- **Validation:** Ensures all adapters are of the same type (SMS, Email, Push)
- **Message type validation:** Ensures all adapters support the same message type
- **Max messages:** Returns the minimum `maxMessagesPerRequest` across all adapters

## Testing

All tests pass:
- ✅ 14 Messenger-specific tests
- ✅ PSR-12 linting
- ✅ PHPStan level 6 static analysis

## Checklist

- [x] Code follows PSR-12 style guidelines
- [x] Static analysis passes (PHPStan level 6)
- [x] Tests added and passing
- [x] README updated with usage example
- [x] No breaking changes to existing adapters

### CODE REVIEW

File: src/Utopia/Messaging/Messenger.php [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                    $e->getMessage()
+                );
+            }
+        }
+
+        throw new \Exception(sprintf(
+            "All %d adapters failed:\n%s",
+            count($this->adapters),
+            implode("\n", $errors)
+        ));
```
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply src/Utopia/Messaging/Messenger.php]: Fixed in 2e823a4. The failure message now pluralizes correctly, so the single-adapter case reads `All 1 adapter failed`.

File: src/Utopia/Messaging/Messenger.php [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        if (empty($adapters)) {
+            throw new \InvalidArgumentException('At least one adapter must be provided.');
+        }
+
+        $this->validateAdapters($adapters);
+
+        $this->adapters = $adapters;
+    }
```
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply src/Utopia/Messaging/Messenger.php:67]: Fixed in 2e823a4. Messenger now validates every array element before compatibility checks, so invalid entries fail with an InvalidArgumentException instead of crashing later with a PHP error. I also added test coverage for the non-Adapter element case.

File: src/Utopia/Messaging/Messenger.php [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                'Invalid message type. Expected "%s", got "%s".',
+                $messageType,
+                get_class($message)
+            ));
+        }
+
+        foreach ($this->adapters as $index => $adapter) {
+            try {
+                return $adapter->send($message);
+            } catch (\Exception $e) {
```
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply src/Utopia/Messaging/Messenger.php:103]: Keeping this one intentional for now. Messenger only fails over on thrown Exceptions, not all Throwables, because catching PHP-level Errors like TypeError would also mask adapter implementation bugs that should stay visible during debugging. If we decide later that failover should also treat those as transport failures, I would rather make that an explicit behavior change.

File: src/Utopia/Messaging/Messenger.php [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+     *
+     * @throws \InvalidArgumentException If adapters are not compatible.
+     */
+    private function validateAdapters(array $adapters): void
+    {
+        $firstAdapter = $adapters[0];
+        $expectedType = $firstAdapter->getType();
+        $expectedMessageType = $firstAdapter->getMessageType();
+
+        foreach ($adapters as $index => $adapter) {
```
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply src/Utopia/Messaging/Messenger.php]: Cleaned up in 2e823a4 while touching validation. The compatibility loop now starts from the second adapter, so we skip the redundant self-check on the first adapter.

**Comment by ChiragAgg5k** [comment]
Closing because this repository is now a read-only mirror. `utopia-php/messaging` moved into [`packages/messaging`](https://github.com/appwrite/appwrite/tree/main/packages/messaging) in appwrite/appwrite (appwrite/appwrite#13863). Every change there is split back here automatically, so a PR merged on the mirror would be overwritten by the next split.

If this change is still wanted, please open it against `packages/messaging` in appwrite/appwrite. The layout is flatter now: sources are under `packages/messaging/src` (for example `src/Adapter/SMS/...`), and tests are under `Utopia\\Messaging\\Tests`, with provider tests in `tests/E2E`. Thanks for the contribution.
